### DISCUSSION TITLE
Best practice for searchbox implementation?

### DISCUSSION THREAD

**Discussion by DiegoMcDipster** [vercel/next.js#20817] [Help] [answered]
Hi,

I'm new to Next js (and to web development).  

I'm building a website that will list available courses. On the Courses page is a card for every course and the school where the course will be held.  You click on the card and get to the Course page.

I want to implement a searchbox to find the courses available at a particular school.

The course data (json) is stored in mongoDB.  I can use getStaticProps to pre-render the pages.  The searchbox however will filter the courses array and I don't think that would work at runtime (?)

The people in this group have much more experience than me, so i'm hoping you can advise on how you would implement it.

Thanks in advance!

**Comment by jamesmosier** [comment, marked as answer]
Hi @DiegoMcDipster. If you have a code sample that'd be super helpful! Otherwise this is one potential way to do it all client side:

```js
export default function MyPage({ courses: initialCourses }) {
  // get the `courses` from `getStaticProps` and store them in state
  const [courses, setCourses] = useState(initialCourses);

  const performFilter = courseName => {
    setCourses(initialCourses.filter(course => course.name === courseName));
  };

  return (
    <div>
      <button type="button" onClick={() => performFilter('course_a')}>
        Course A
      </button>
      
      {courses.map(course => (
        <div key={course.name}>{course.name}</div>
      ))}
    </div>
  );
}
```

**Reply by DiegoMcDipster** [reply]
Thanks @jamesmosier for your assistance!

I didn't post any code because there isn't any yet... I'm still trying to get my head round how getStaticProps works. What I've understood about getStaticProps is that it only runs at buildtime (note, I'm not using the revalidate option).

My understanding of what you are saying is that:
1. build time the getStaticProps will return the courses and I can save the courses in state
2. At runtime the courses array in state will be available and the filtering will work 

Did I understand that correctly?

**Reply by jamesmosier** [reply]
Yep, you got it! Then you'll be able to display the courses immediately (since they were fetched from the database at build time) but then you can have the interactivity at runtime and use React state to manage what to show.

**Reply by DiegoMcDipster** [reply]
Brilliant!  Thanks.  I'll start building it!

**Reply by djvid** [reply]
> Hi @DiegoMcDipster. If you have a code sample that'd be super helpful! Otherwise this is one potential way to do it all client side:
> 
> ```js
> export default function MyPage({ courses: initialCourses }) {
>   // get the `courses` from `getStaticProps` and store them in state
>   const [courses, setCourses] = useState(initialCourses);
> 
>   const performFilter = courseName => {
>     setCourses(initialCourses.filter(course => course.name === courseName));
>   };
> 
>   return (
>     <div>
>       <button type="button" onClick={() => performFilter('course_a')}>
>         Course A
>       </button>
>       
>       {courses.map(course => (
>         <div key={course.name}>{course.name}</div>
>       ))}
>     </div>
>   );
> }
> ```

Hey @jamesmosier, sorry to hijack this old discussion but I have a similar problem. I have an e-commerce site where users can filter over a thousand products on different variants like size, color etc and since I want to use getStaticProps I can't do it the traditional way using query strings. Is it good good practice performance wise, to query all 1k+ products and do the filtering/sorting/pagination on the client side?
